File: src/ReverseLinkedList.java
/*
 * Title: Reverse A Linked List
 * 
 * Problem: Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zero elements.
 * 
 * Author: Veeral Suthar
 *
 * Date: 07/24/2022
 * 
 * Solution: Iterative and recursive solution
 * Time Complexity: Iterative: Recursive: 
 */

public class ReverseLinkedList {

    public class ListNode {
        int val;
        ListNode next;
         ListNode() {}
         ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode iterativelyReverseList(ListNode head) {
        ListNode prev = null;
        
        while (head != null) {
            ListNode next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }

    public ListNode recursivelyReverseList(ListNode head) {
        return reverse(head,null);
    } 
    private ListNode reverse(ListNode head, ListNode newHead) {
        if (head == null) return newHead;
        ListNode next = head.next;
        head.next = newHead;
        return reverse(next, head);
    }
    
}

File: Swift/ReverseLinkedList.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class ReverseLinkedList {

    // Iteratively reverse the linked list
    func reverseList(_ head: ListNode?) -> ListNode? {
        var prev: ListNode? = nil
        var curr: ListNode? = head
        while curr != nil {
            let temp: ListNode? = curr?.next
            curr?.next = prev
            prev = curr
            curr = temp
        }
        return prev
    }

    // Recusively reverse the linked list
    func recursivelyReverseList(_ head: ListNode?) -> ListNode? {
        
        // Empty List
	    guard let head = head else { return nil }
        // Base Case
	    guard let next = head.next else { return head }

	    let new = recursivelyReverseList(next)
	    next.next = head
	    head.next = nil

	    return new   
    }
}
